ext/opentelemetry-ext-sqlalchemy/src/opentelemetry/ext/sqlalchemy/engine.py

```diff
@@ -0,0 +1,162 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from sqlalchemy.event import listen
+
+from opentelemetry import trace
+from opentelemetry.ext.sqlalchemy.version import __version__
+from opentelemetry.trace.status import Status, StatusCanonicalCode
+
+# Network attribute semantic convention here:
+# https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/span-general.md#general-network-connection-attributes
+_HOST = "net.peer.name"
+_PORT = "net.peer.port"
+# Database semantic conventions here:
+# https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/database.md
+_ROWS = "sql.rows"  # number of rows returned by a query
+_STMT = "db.statement"
+_DB = "db.type"
+_URL = "db.url"
+
+
+def _normalize_vendor(vendor):
+    """ Return a canonical name for a type of database. """
+    if not vendor:
+        return "db"  # should this ever happen?
+
+    if "sqlite" in vendor:
+        return "sqlite"
+
+    if "postgres" in vendor or vendor == "psycopg2":
+        return "postgres"
+
+    return vendor
+
+
+def _get_tracer(engine, tracer_provider=None):
+    tracer = getattr(engine, "_opentelemetry_tracer", False)
+    if tracer:
```

Comment:
Is there ever a case when the `TracerProvider` is switched during runtime? Wouldn't the tracer that is returned after that always correspond to the first one that was created?

Reply:
Good catch, storing the attr on the engine is no longer needed so i've removed this.